File: pkg/geometry/cylinder.go
package geometry

import (
	"math"

	"sif/gogs/eric/relray/pkg/vec"
)

// Cylinder is a capped cylinder along the local Y axis.
// It extends from Y=0 to Y=Height, centered at X=0, Z=0.
// Use a Transformed wrapper to position and orient it in the scene.
type Cylinder struct {
	Radius float64
	Height float64
}

func (c *Cylinder) Intersect(origin, dir vec.Vec3, tMin, tMax float64) (Hit, bool) {
	// Solve for intersection with infinite cylinder x² + z² = r²
	a := dir.X*dir.X + dir.Z*dir.Z
	b := 2 * (origin.X*dir.X + origin.Z*dir.Z)
	cc := origin.X*origin.X + origin.Z*origin.Z - c.Radius*c.Radius

	var bestT float64 = -1
	var bestNormal vec.Vec3
	found := false

	disc := b*b - 4*a*cc
	if disc >= 0 && a > 1e-12 {
		sqrtDisc := math.Sqrt(disc)
		for _, t := range [2]float64{(-b - sqrtDisc) / (2 * a), (-b + sqrtDisc) / (2 * a)} {
			if t < tMin || t > tMax {
				continue
			}
			y := origin.Y + t*dir.Y
			if y >= 0 && y <= c.Height {
				if !found || t < bestT {
					bestT = t
					p := origin.Add(dir.Scale(t))
					bestNormal = vec.Vec3{X: p.X, Z: p.Z}.Normalize()
					found = true
				}
				break // first valid t from sorted pair is closest
			}
		}
	}

	// Test caps (Y=0 and Y=Height)
	if math.Abs(dir.Y) > 1e-12 {
		for _, capY := range [2]float64{0, c.Height} {
			t := (capY - origin.Y) / dir.Y
			if t < tMin || t > tMax {
				continue
			}
			p := origin.Add(dir.Scale(t))
			if p.X*p.X+p.Z*p.Z <= c.Radius*c.Radius {
				if !found || t < bestT {
					bestT = t
					if capY == 0 {
						bestNormal = vec.Vec3{Y: -1}
					} else {
						bestNormal = vec.Vec3{Y: 1}
					}
					found = true
				}
			}
		}
	}

	if !found {
		return Hit{}, false
	}
	p := origin.Add(dir.Scale(bestT))
	h := Hit{T: bestT, Point: p}
	h.SetFaceNormal(dir, bestNormal)
	return h, true
}

File: rust/src/geometry/cylinder.rs
use crate::vec::Vec3;
use super::{Hit, Shape};

pub struct Cylinder {
    pub radius: f64,
    pub height: f64,
}

impl Shape for Cylinder {
    fn intersect(&self, origin: Vec3, dir: Vec3, t_min: f64, t_max: f64) -> Option<Hit> {
        let a = dir.x * dir.x + dir.z * dir.z;
        let b = 2.0 * (origin.x * dir.x + origin.z * dir.z);
        let cc = origin.x * origin.x + origin.z * origin.z - self.radius * self.radius;

        let mut best_t: f64 = -1.0;
        let mut best_normal = Vec3::default();
        let mut found = false;

        let disc = b * b - 4.0 * a * cc;
        if disc >= 0.0 && a > 1e-12 {
            let sqrt_disc = disc.sqrt();
            for t in [(-b - sqrt_disc) / (2.0 * a), (-b + sqrt_disc) / (2.0 * a)] {
                if t < t_min || t > t_max {
                    continue;
                }
                let y = origin.y + t * dir.y;
                if y >= 0.0 && y <= self.height {
                    if !found || t < best_t {
                        best_t = t;
                        let p = origin + dir * t;
                        best_normal = Vec3 { x: p.x, y: 0.0, z: p.z }.normalize();
                        found = true;
                    }
                    break;
                }
            }
        }

        // Test caps (Y=0 and Y=Height)
        if dir.y.abs() > 1e-12 {
            for cap_y in [0.0, self.height] {
                let t = (cap_y - origin.y) / dir.y;
                if t < t_min || t > t_max {
                    continue;
                }
                let p = origin + dir * t;
                if p.x * p.x + p.z * p.z <= self.radius * self.radius {
                    if !found || t < best_t {
                        best_t = t;
                        best_normal = if cap_y == 0.0 {
                            Vec3 { x: 0.0, y: -1.0, z: 0.0 }
                        } else {
                            Vec3 { x: 0.0, y: 1.0, z: 0.0 }
                        };
                        found = true;
                    }
                }
            }
        }

        if !found {
            return None;
        }
        let p = origin + dir * best_t;
        let mut h = Hit { t: best_t, point: p, ..Default::default() };
        h.set_face_normal(dir, best_normal);
        Some(h)
    }
}
